core/go/internal/sequencer/coordinator/coordinator.go

```diff
@@ -0,0 +1,600 @@
+					return
+				}
+
+				// There are too many transactions making their way to the base ledger. If we're hitting this frequently
+				// then the node has probably been configured to optimise for base-ledger safety rather than optimal
+				// throughput
+				time.Sleep(100 * time.Millisecond)
+			}
+			log.L(ctx).Debugf("coordinator submitting transaction %s for dispatch", tx.ID.String())
+			c.readyForDispatch(ctx, tx)
+		case <-c.stopDispatchLoop:
+			log.L(ctx).Infof("coordinator dispatch loop stopped")
+			return
+		case <-ctx.Done():
+			log.L(ctx).Infof("coordinator dispatch loop cancelled")
+			return
+		}
+	}
+}
+
+func (c *coordinator) sendHandoverRequest(ctx context.Context) {
+	err := c.transportWriter.SendHandoverRequest(ctx, c.activeCoordinatorNode, c.contractAddress)
+	if err != nil {
+		log.L(ctx).Errorf("error sending handover request: %v", err)
+	}
+}
+
+func (c *coordinator) GetActiveCoordinatorNode(ctx context.Context) string {
+	if c.activeCoordinatorNode == "" {
+		// If we don't yet have an active coordinator, select one based on the appropriate algorithm for the contract type
+		activeCoordinator, err := c.SelectActiveCoordinatorNode(ctx)
+		if err != nil {
+			log.L(ctx).Errorf("error selecting next active coordinator: %v", err)
+			return ""
+		}
+		c.activeCoordinatorNode = activeCoordinator
+	}
+	return c.activeCoordinatorNode
+}
+
+func (c *coordinator) SelectActiveCoordinatorNode(ctx context.Context) (string, error) {
+	coordinator := ""
+	if c.domainAPI.ContractConfig().GetCoordinatorSelection() == prototk.ContractConfig_COORDINATOR_STATIC {
+		// E.g. Noto
+		if c.domainAPI.ContractConfig().GetStaticCoordinator() == "" {
+			return "", fmt.Errorf("static coordinator mode is configured but static coordinator node is not set")
+		}
+		log.L(ctx).Debugf("coordinator %s selected as next active coordinator in static coordinator mode", c.domainAPI.ContractConfig().GetStaticCoordinator())
+		// If the static coordinator returns a fully qualified identity extract just the node name
+		if strings.Contains(c.domainAPI.ContractConfig().GetStaticCoordinator(), "@") {
+			parts := strings.Split(c.domainAPI.ContractConfig().GetStaticCoordinator(), "@")
+			coordinator = parts[1]
+		} else {
+			coordinator = c.domainAPI.ContractConfig().GetStaticCoordinator()
+		}
+	} else if c.domainAPI.ContractConfig().GetCoordinatorSelection() == prototk.ContractConfig_COORDINATOR_ENDORSER {
+		// E.g. Pente
+		// Make a fair choice about the next coordinator, but for now just choose the node this request arrived at
+		if len(c.originatorNodePool) == 0 {
+			log.L(ctx).Warnf("no pool to select a coordinator from yet")
+			return "", nil
+		} else {
+			// Round block number down to the nearest block range (e.g. block 1012, 1013, 1014 etc. all become 1000 for hashing)
+			effectiveBlockNumber := c.currentBlockHeight - (c.currentBlockHeight % c.coordinatorSelectionBlockRange)
```

Comment:
Interesting in learning how inactive coordinators get evicted from the `originatorNodePool` in N of M HA scenarios